[FLINK-6764] Deduplicate stateless serializers in checkpoints

tzulitai opened the pull request:
This PR is based on #4014, so only the last commit 39ffe7e is relevant.

Prior to this PR, we would write multiple instances of the same serializer even if it was stateless. This commit changes that by first writing a serializer index at the head of the stream, and only write the index of a serializer when one needs to be written. The index map is built using `IdentitiyHashMap`s, so that stateful serializers are considered as separate entries in the index.

## Test

New tests are added to `PojoSerializerTest` and `SerializationProxiesTest` to test that stateless serializers are not duplicated on restore.

StefanRRichter commented on flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializerTest.java:
```diff
@@ -606,6 +665,45 @@ public void testSerializerSerializationFailureResilience() throws Exception {
+		PojoSerializer.PojoSerializerConfigSnapshot<TestUserClassWithSameFieldTypes> config = pojoSerializer.snapshotConfiguration();
+		byte[] serializedConfig;
+		try (
+			ByteArrayOutputStream out = new ByteArrayOutputStream()) {
```
You could use Flink's `ByteArrayOutputStreamWithPos`, which can be more efficient because it does not use `synchronized` and you can access the write position and the internal array to immediately write it to the stream instead of creating a copy in the `out.toByteArray()` part.

StefanRRichter commented on flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializerTest.java:
```diff
@@ -606,6 +665,45 @@ public void testSerializerSerializationFailureResilience() throws Exception {
+
+		// read configuration from bytes
+		PojoSerializer.PojoSerializerConfigSnapshot<TestUserClass> deserializedConfig;
+		try(ByteArrayInputStream in = new ByteArrayInputStream(serializedConfig)) {
```
There is also a `ByteArrayInputStreamWithPos`.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:
```diff
@@ -795,13 +782,15 @@ public PojoSerializerConfigSnapshot(
+
 			// --- write fields and their serializers, in order
 
 			out.writeInt(fieldToSerializerConfigSnapshot.size());
```
What I just noticed is that this is almost 3x duplicated code. I suggest to create a private method that writes `Map<Class<?>, Tuple2<TypeSerializer<?>, TypeSerializerConfigSnapshot>>` and call it 3 times.

StefanRRichter replied:
The same might hold for reads.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeTypeSerializerConfigSnapshot.java:
```diff
@@ -110,4 +123,20 @@ public boolean equals(Object obj) {
 		return nestedSerializersAndConfigs.hashCode();
 	}
+
+	private Map<TypeSerializer<?>, Integer> buildSerializerIndices() {
```
While the idea makes sense, I am not sure if we can rely on all serializer having a correct implementation of equals and hash code that will do what we want. We might be better of with collection of `serializer -> index` pairs and using equality for deduplication, but even there equals might be tricky. While it is ok to accidentally write serializer twice, it should never happen that a serializer gets de-duplicated by accident.

StefanRRichter replied:
I see that the concrete implementation is always an identity hash map. since it behaves different from what you expect by a normal map (using equals/hashcode) and in that sense violates LSP, I suggest you wrap it in an own class called `SerializerIndex`.

StefanRRichter commented:
Overall, very nice work! I just had few minor comments. After they are addressed, this is good to merge. +1

tzulitai commented:
Thanks a lot for the helpful review @StefanRRichter. I've addressed your comments.

This PR also needs one more additional change for backwards compatibility with 1.3.0 before I'll merge it (will ping you on the follow-up).

zentol commented:
@tzulitai What's the state of this PR?

tzulitai commented:
@zentol there was a recent offline discussion regarding the savepoint formats that will probably void this change. Will close this PR for now.
